Add heap infrastructure and simple ptmalloc model

tgduckworth opened the pull request:
# Overview

I took a shot at improving the heap situation in angr. This pull request does not change any of angr's default behavior, so if you merge this you should see no change in the behavior of the current heap model. The changes do introduce more infrastructure around the heap, making it more accessible to users and introducing the ability to switch heap models. As an example, I've implemented a very very *very* simple model of ptmalloc (although due to its limitations it's much more like dlmalloc right now). A `heap` state plugin has been added, and users can modify the state while performing dynamic memory management with functions like `state.heap.malloc()` and `state.heap.free()`.

This attempts to address #762.

# Interfacing with angr

angr has `SimProcedure`s for libc's `malloc`, `free`, `calloc`, and `realloc`. Under different heap implementations, these should affect the program state differently. To this end, these `SimProcedure`s are no longer directly responsible for this functionality. They now defer requests made of them to a heap plugin, which may be one of many implementations.

To be faithful to angr's current expectations of the `SimProcedure`s listed above, all heaps are expected to implement some kind of behavior for these libc functions via `_malloc`, `_free`, `_calloc`, and `_realloc` heap methods. Heaps that actually support (near-)standard libc functionality should alias these to a public method as well. As an example, the `SimHeapPTMalloc` class makes these functions public since it approximates the expected behavior of these functions. For the current angr model (which I've moved to `SimHeapBrk`), these are not made public. Instead, it offers `allocate()` and `release()` as a sort of "push" and "pop" model (although it's important to note that, in the interest in preserving the old functionality, the `free` `SimProcedure` does NOT end up using the `release()` functionality in any way).

# Important Files

Most of the relevant code can be found in the `angr.state_plugins.heap` module. A potential class hierarchy might be envisioned to look something like the diagram seen below. BiBOP heaps are "big bags of pages". They differ from the classic freelist-style heap in that they allocate entire pages that are used to store objects of the same size. From what I've read, jemalloc and the Windows 10 heap are implemented in this fashion. On the other hand, ptmalloc (derived from dlmalloc, and used in glibc) and the classic Windows heaps are freelist heap implementations. I call angr's current heap implementation a brk heap since it just moves the brk up and up forever. Real heaps do manipulate the brk obviously but it is not the basis for the heap management, hence the earlier relation to a stack.

```plain
                   ___BaseHeap___
                  /      |       \
         BiBOPHeap    brkHeap     FreeListHeap
        /       |                 |          \
jemalloc      W10Heap           ptmalloc      NTHeap
```

In the aforementioned module you will find a few files corresponding to the different kinds of heaps considered right now (not all of the above heaps and heap types are implemented right now).
* `SimHeapBase`: The basic interface for all heaps. Specifies basic heap properties and expected functionality.
* `SimHeapBrk`: This is where the old angr heap functionality now lives. Whenever a `SimState` is created without an explicit heap specified, this is the default heap plugin that gets attached.
* `SimHeapFreeList`: The basic interface for all freelist heaps. It also defines the notion of a chunk in such a heap. It may be a bit biased towards the dlmalloc style.
* `SimHeapLibc`: This is meant to be a superclass of any heap implementing (near-)standard libc functionality. It just specifies that the libc functions should be publicly exposed for ease-of-use. It sits in the class hierarchy just below `SimHeapBase` at the moment and is subclassed by `SimHeapFreeList`, but should probably be a mixin or whatever if someone wants to investigate that further.
* `SimHeapPTMalloc`: This is a very basic interpretation of ptmalloc that looks a lot more like a very basic interpretation of dlmalloc. It's more efficient than the old angr functionality for memory management purposes and may offer greater fidelity if you're looking for basic heap bugs, but it's not *that* sophisticated.

# Considerations for `SimHeapPTMalloc`

This heap makes all sorts of concretization assumptions because the path explosion would be quite substantial if certain pieces of metadata became symbolic. Whenever such a concretization is made, a warning is issued. As an example, if the metadata for a chunk's size is overwritten with symbolic data, a whole bunch of new paths are possible for symbolic execution, many of which become chaotic quickly since a user assigning an arbitrary size to a chunk can have all sorts of effects on the rest of the heap state after the fact. Having the bit controlling a chunk's allocation or freedom become symbolic has similar implications since that one bit governs the presence or absence of other fields, which can impact the legality of other heap operations, cause other symbolic data to become relevant metadata, and so on.

A special case of such a concretization strategy can be seen [here](https://github.com/tgduckworth/angr/blob/feat/heap/master/angr/state_plugins/heap/heap_base.py#L33) with some more in-depth discussion. Other instances of concretization assumptions could probably be pursued after a satisfactory solution to this one is found, but for now they just issue warnings when a decision must be made, and then they concretize in a manner I assumed would be sane to start with. In the future there could probably be better user control over these decisions.

`SimHeapPTMalloc` is not capable of requesting additional memory from the system at the moment, so it simply manages the memory that was given to it to start with.

# Example Usage

```python
import angr
import logging

p = angr.Project('/bin/true', auto_load_libs=False)

# To create a state with a `SimHeapBrk` heap, with the old angr behavior
s = p.factory.entry_state()

# Or explicitly
s = p.factory.entry_state(heap=angr.SimHeapBrk())

# Or explicitly with params (note: `SimHeapBrk` does not respect its size)
s = p.factory.entry_state(heap=angr.SimHeapBrk(heap_base=0xd0000000, heap_size=32 * 4096))

# To create a state with a `SimHeapPTMalloc` heap, with optional params
s = p.factory.entry_state(heap=angr.SimHeapPTMalloc(heap_base=0xd0000000, heap_size=32 * 4096))

# As usual, you can do plugin registration manually
s.register_plugin('heap', angr.SimHeapPTMalloc(heap_base=0xd0000000, heap_size=0x1000))

s.heap.print_all_chunks()
# Freelist heaps support the above, which produces:
# ┌───────────────────────────────┐
# ├───────── HEAP CHUNKS ─────────┤
# │ <PTChunk (free @ 0xd0000000)> │
# └───────────────────────────────┘

# Interspersing the following statements with prints and the heap's debug statements, you'd get this trace on amd64
logging.getLogger('angr.state_plugins.heap').setLevel('DEBUG')

s.heap.malloc(100)
s.heap.print_all_chunks()

# DEBUG   | 2019-02-04 12:36:44,955 | angr.state_plugins.heap.heap_ptmalloc | Requested:  100; Allocated:  128; Returned: 0xd0000010; Chunk: 0xd0000000
# ┌───────────────────────────────┐
# ├───────── HEAP CHUNKS ─────────┤
# │ <PTChunk (used @ 0xd0000000)> │
# │ <PTChunk (free @ 0xd0000080)> │
# └───────────────────────────────┘

a1 = s.heap.calloc(7, 32)
s.heap.print_all_chunks()

# DEBUG   | 2019-02-04 12:36:45,022 | angr.state_plugins.heap.heap_ptmalloc | Requested:  224; Allocated:  240; Returned: 0xd0000090; Chunk: 0xd0000080
# ┌───────────────────────────────┐
# ├───────── HEAP CHUNKS ─────────┤
# │ <PTChunk (used @ 0xd0000000)> │
# │ <PTChunk (used @ 0xd0000080)> │
# │ <PTChunk (free @ 0xd0000170)> │
# └───────────────────────────────┘

s.heap.malloc(5)
s.heap.print_all_chunks()

# DEBUG   | 2019-02-04 12:36:45,139 | angr.state_plugins.heap.heap_ptmalloc | Requested:    5; Allocated:   32; Returned: 0xd0000180; Chunk: 0xd0000170
# ┌───────────────────────────────┐
# ├───────── HEAP CHUNKS ─────────┤
# │ <PTChunk (used @ 0xd0000000)> │
# │ <PTChunk (used @ 0xd0000080)> │
# │ <PTChunk (used @ 0xd0000170)> │
# │ <PTChunk (free @ 0xd0000190)> │
# └───────────────────────────────┘

a2 = s.heap.malloc(0)
s.heap.print_all_chunks()

# DEBUG   | 2019-02-04 12:36:45,265 | angr.state_plugins.heap.heap_ptmalloc | Requested:    0; Allocated:   32; Returned: 0xd00001a0; Chunk: 0xd0000190
# ┌───────────────────────────────┐
# ├───────── HEAP CHUNKS ─────────┤
# │ <PTChunk (used @ 0xd0000000)> │
# │ <PTChunk (used @ 0xd0000080)> │
# │ <PTChunk (used @ 0xd0000170)> │
# │ <PTChunk (used @ 0xd0000190)> │
# │ <PTChunk (free @ 0xd00001b0)> │
# └───────────────────────────────┘

a3 = s.heap.malloc(643)
s.heap.print_all_chunks()

# DEBUG   | 2019-02-04 12:36:45,363 | angr.state_plugins.heap.heap_ptmalloc | Requested:  643; Allocated:  672; Returned: 0xd00001c0; Chunk: 0xd00001b0
# ┌───────────────────────────────┐
# ├───────── HEAP CHUNKS ─────────┤
# │ <PTChunk (used @ 0xd0000000)> │
# │ <PTChunk (used @ 0xd0000080)> │
# │ <PTChunk (used @ 0xd0000170)> │
# │ <PTChunk (used @ 0xd0000190)> │
# │ <PTChunk (used @ 0xd00001b0)> │
# │ <PTChunk (free @ 0xd0000450)> │
# └───────────────────────────────┘

a4 = s.heap.malloc(25)
s.heap.print_all_chunks()

# DEBUG   | 2019-02-04 12:36:45,476 | angr.state_plugins.heap.heap_ptmalloc | Requested:   25; Allocated:   48; Returned: 0xd0000460; Chunk: 0xd0000450
# ┌───────────────────────────────┐
# ├───────── HEAP CHUNKS ─────────┤
# │ <PTChunk (used @ 0xd0000000)> │
# │ <PTChunk (used @ 0xd0000080)> │
# │ <PTChunk (used @ 0xd0000170)> │
# │ <PTChunk (used @ 0xd0000190)> │
# │ <PTChunk (used @ 0xd00001b0)> │
# │ <PTChunk (used @ 0xd0000450)> │
# │ <PTChunk (free @ 0xd0000480)> │
# └───────────────────────────────┘

s.heap.free(a1)
s.heap.print_all_chunks()

# DEBUG   | 2019-02-04 12:36:45,548 | angr.state_plugins.heap.heap_ptmalloc | Free request: 0xd0000090; Freed chunk: 0xd0000080
# ┌───────────────────────────────┐
# ├───────── HEAP CHUNKS ─────────┤
# │ <PTChunk (used @ 0xd0000000)> │
# │ <PTChunk (free @ 0xd0000080)> │
# │ <PTChunk (used @ 0xd0000170)> │
# │ <PTChunk (used @ 0xd0000190)> │
# │ <PTChunk (used @ 0xd00001b0)> │
# │ <PTChunk (used @ 0xd0000450)> │
# │ <PTChunk (free @ 0xd0000480)> │
# └───────────────────────────────┘

a5 = s.heap.malloc(20)
s.heap.print_all_chunks()

# DEBUG   | 2019-02-04 12:36:45,680 | angr.state_plugins.heap.heap_ptmalloc | Requested:   20; Allocated:   48; Returned: 0xd0000090; Chunk: 0xd0000080
# ┌───────────────────────────────┐
# ├───────── HEAP CHUNKS ─────────┤
# │ <PTChunk (used @ 0xd0000000)> │
# │ <PTChunk (used @ 0xd0000080)> │
# │ <PTChunk (free @ 0xd00000b0)> │
# │ <PTChunk (used @ 0xd0000170)> │
# │ <PTChunk (used @ 0xd0000190)> │
# │ <PTChunk (used @ 0xd00001b0)> │
# │ <PTChunk (used @ 0xd0000450)> │
# │ <PTChunk (free @ 0xd0000480)> │
# └───────────────────────────────┘

s.heap.free(a2)
s.heap.print_all_chunks()

# DEBUG   | 2019-02-04 12:36:45,774 | angr.state_plugins.heap.heap_ptmalloc | Free request: 0xd00001a0; Freed chunk: 0xd0000190
# ┌───────────────────────────────┐
# ├───────── HEAP CHUNKS ─────────┤
# │ <PTChunk (used @ 0xd0000000)> │
# │ <PTChunk (used @ 0xd0000080)> │
# │ <PTChunk (free @ 0xd00000b0)> │
# │ <PTChunk (used @ 0xd0000170)> │
# │ <PTChunk (free @ 0xd0000190)> │
# │ <PTChunk (used @ 0xd00001b0)> │
# │ <PTChunk (used @ 0xd0000450)> │
# │ <PTChunk (free @ 0xd0000480)> │
# └───────────────────────────────┘

s.heap.free(a4)
s.heap.print_all_chunks()

# DEBUG   | 2019-02-04 12:36:45,856 | angr.state_plugins.heap.heap_ptmalloc | Free request: 0xd0000460; Freed chunk: 0xd0000450
# ┌───────────────────────────────┐
# ├───────── HEAP CHUNKS ─────────┤
# │ <PTChunk (used @ 0xd0000000)> │
# │ <PTChunk (used @ 0xd0000080)> │
# │ <PTChunk (free @ 0xd00000b0)> │
# │ <PTChunk (used @ 0xd0000170)> │
# │ <PTChunk (free @ 0xd0000190)> │
# │ <PTChunk (used @ 0xd00001b0)> │
# │ <PTChunk (free @ 0xd0000450)> │
# └───────────────────────────────┘

s.heap.free(a3)
s.heap.print_all_chunks()

# DEBUG   | 2019-02-04 12:36:45,923 | angr.state_plugins.heap.heap_ptmalloc | Free request: 0xd00001c0; Freed chunk: 0xd00001b0
# ┌───────────────────────────────┐
# ├───────── HEAP CHUNKS ─────────┤
# │ <PTChunk (used @ 0xd0000000)> │
# │ <PTChunk (used @ 0xd0000080)> │
# │ <PTChunk (free @ 0xd00000b0)> │
# │ <PTChunk (used @ 0xd0000170)> │
# │ <PTChunk (free @ 0xd0000190)> │
# └───────────────────────────────┘

s.heap.realloc(a5, 200)
s.heap.print_all_chunks()

# DEBUG   | 2019-02-04 12:36:45,999 | angr.state_plugins.heap.heap_ptmalloc | Requested:  200; Allocated:  224; Returned: 0xd00001a0; Chunk: 0xd0000190
# DEBUG   | 2019-02-04 12:36:46,039 | angr.state_plugins.heap.heap_ptmalloc | Free request: 0xd0000090; Freed chunk: 0xd0000080
# ┌───────────────────────────────┐
# ├───────── HEAP CHUNKS ─────────┤
# │ <PTChunk (used @ 0xd0000000)> │
# │ <PTChunk (free @ 0xd0000080)> │
# │ <PTChunk (used @ 0xd0000170)> │
# │ <PTChunk (used @ 0xd0000190)> │
# │ <PTChunk (free @ 0xd0000270)> │
# └───────────────────────────────┘
```

rhelmot commented on angr/state_plugins/heap/heap_base.py:
```diff
@@ -0,0 +1,118 @@
+        raise NotImplementedError("%s not implemented for %s" % (self._free.__func__.__name__,
+                                                                 self.__class__.__name__))
+
+    def _calloc(self, sim_nmemb, sim_size):
```
Why is there an explicit calloc? Is there any benefit to having calloc done internally instead of pushing it out to the simprocedure which calls malloc?

tgduckworth replied:
Good question. At the time I only did it to be consistent with the treatment the other functions were getting.

If I'm understanding correctly, you're asking whether it would be possible to implement `calloc` in a common manner for the heaps we have right now (by implementing it in terms of `malloc`). I'm pretty sure that would be fine right now.

m1ghtym0 replied:
We might want to check that `mmap_base` is not overlapping with the heap:
`self.mmap_base = DEFAULT_HEAP_LOCATION + DEFAULT_HEAP_SIZE * 2` 
Since heap_location, was set to the default location, this was implicitly assumed so far.

tgduckworth replied:
Good catch. Probably deserves better treatment than [this](https://github.com/angr/angr/pull/1432/commits/c8a856677f2dc9fe83ae2f9f24dafe629f1515a5) in the long run, but it should be a bit more sane with respect to the possibility you bring up.

rhelmot commented on angr/procedures/libc/malloc.py:
```diff
@@ -14,14 +14,4 @@ class malloc(angr.SimProcedure):
-        addr = self.state.libc.heap_location
-        self.state.libc.heap_location += size
-        return addr
+        return self.state.heap._malloc(sim_size)
```
Why should this simprocedure call `_malloc` instead of `malloc`? What's an example of a plugin that would not export a `malloc`?

rhelmot replied:
I see that the brk heap doesn't have a malloc, but instead has an allocate... I worry about the interface fragmentation, but I'm tentatively okay with this for now.

tgduckworth replied:
My expectation is that most (if not all) plugins will export `malloc`, but yes, I was concerned about making it and `free` available publicly for the default heap since that one might not behave at all in the way a casual user would expect. On the other hand, it's just exposing the old functionality of the `SimProcedure`s, so maybe that isn't as big a deal as I had considered it.

rhelmot commented on angr/state_plugins/heap/utils.py:
```diff
@@ -0,0 +1,19 @@
+    :returns: a concrete value for the item
+    """
+    if solver.symbolic(x):
+        sols = solver.eval_upto(x, 2)
```
The kosher way of doing this these days is `state.solver.eval_one(x)` and then catching the SimSolverError if there's more than one

tgduckworth replied:
Addressed [here](https://github.com/angr/angr/pull/1432/commits/1ed93117837470dcb9460f1bb8d109a7d1675b50).

rhelmot commented on tests/test_ptmalloc.py:
```diff
@@ -0,0 +1,282 @@
+
+def test_calloc_clears():
+    for arch in ('X86', 'AMD64'):
+        yield run_calloc_clears, arch
```
Could you add an `if __name__ == '__main__':` clause that runs all the tests?

tgduckworth replied:
Addressed [here](https://github.com/angr/angr/pull/1432/commits/4b2eb577bfee89b3f0c56853ea0ecdd5539b5650).

rhelmot commented on angr/sim_state.py:
```diff
@@ -52,8 +52,9 @@ class SimState(PluginHub):
-                 add_options=None, remove_options=None, special_memory_filler=None, os_name=None, plugin_preset='default', **kwargs):
+    def __init__(self, project=None, arch=None, plugins=None, memory_backer=None, permissions_backer=None, mode=None,
+                 options=None, add_options=None, remove_options=None, special_memory_filler=None, os_name=None,
+                 heap=None, plugin_preset='default', **kwargs):
```
do we want to have heap broken out into its own kwarg or to have it be part of the plugins? We can have `heap=xx` on the SimOS state factory level.

tgduckworth replied:
I think that would be reasonable. The only caveat is that, because of the dependency of the heap plugin on the memory plugin, doing so will add some minor complexity to `SimState`'s `__init__`. Something like this.

```python
        if plugins is not None:
            for n,p in plugins.items():
                if n == 'heap':
                    continue
                self.register_plugin(n, p, inhibit_init=True)
            for p in plugins.values():
                if isinstance(p, angr.state_plugins.heap.SimHeapBase):
                    continue
                p.init_state()
```

And then at the end of `__init__`, we add the following.

```python
        if plugins is not None:
            h = plugins['heap']
            if h is not None:
                self.register_plugin('heap', h)
```

That's not too bad though, and if it is, maybe there's a more elegant way of doing it that I'm ignorant of at the moment. Let me know what you think and I can add the above if this is the direction you want to go in.

rhelmot replied:
So! We have a very janky tool to deal with dependencies between plugins - this is the difference between the `set_state` and `init_state` methods. `set_state` is where you need to do any initialization that doesn't depend on other plugins, and then `init_state` is where you depend on other plugins. This doesn't work if you have any code in `init_state` which depends on another plugin's `init_state` having already run, but the memory plugin has no `init_state`, so this shouldn't be a problem at all!

tgduckworth replied:
So this has brought me back to some problems I had previously while trying to get the plugin initialized at the right time. If I understand things correctly, the `memory` plugin [doesn't actually exist when the rest of the plugins are initialized](https://github.com/angr/angr/blob/master/angr/sim_state.py#L91)--`sym_memory`, `fast_memory`, and `abs_memory` do, but from what I can tell these would be both dangerous to use directly and kind of messy. `SimState` goes on to create the `memory` plugin afterwards depending on what options have been specified, and in my usage I have not seen the args to it be anything other than `None`. I could totally have missed the channel by which the backing memory for `sym_memory` gets passed on to `memory`, but in my experimentation it seems like I end up with differing views of memory depending on which I use, and when. As it is now, I cannot refer to the state's `memory` plugin from the heap plugin's `init_state` without getting an attribute error.

If I move the plugin initialization to be after the `memory` plugin is set up, things work fine, except for the `SimState`'s `_global_condition` attribute [not having been set yet](https://github.com/angr/angr/blob/master/angr/sim_state.py#L146). If I move that up as well, to be set before the plugins are initialized, then things appear to work as they should.

That being said, I feel like I'm playing with fire since `SimState`'s `__init__` is pretty important and attaches to things I'm not too familiar with. Can you verify whether I'm off track here?

tgduckworth replied:
Forgot to add: the nonexistent `_global_condition` gets referenced by the `memory` plugin when stores are requested during a heap's `init_state`.

m1ghtym0 commented:
I've adjusted heaphopper in branch feat/heap, but it's still breaking something I haven't figured out yet.
Chunks get allocated in weird places, but the posix plugin didn't change, so I'm confused.

m1ghtym0 commented:
If you add this, heaphopper is green on the feat/heap branch, which I'll merge as soon as this is merged.

m1ghtym0 commented on angr/state_plugins/heap/heap_brk.py:
```diff
@@ -0,0 +1,128 @@
+    @SimStatePlugin.memo
+    def copy(self, memo):# pylint: disable=unused-argument
+        c = SimHeapBrk()
+        c.heap_location = self.heap_location
```
`SimHeapBrk(heap_base=self.heap_base, heap_size=self.heap_size)`

tgduckworth replied:
Sorry about that! Should be [fixed](https://github.com/angr/angr/pull/1432/commits/cc7a129f3831b137e698c6365171e6ab316d98a5).

m1ghtym0 replied:
No worries! Thanks!

tgduckworth commented:
Hi all. Just want to check in and make sure everything is in good standing with this pull request. I understand that it might be on standby while greater forces are at work, but I'm also a little unsure as to whether the ball is in my court right now or not. I think I've addressed most of the issues, the exception being the plugin initialization, in which case I'm unclear on the best direction to take.

In any case, please feel free to bug me about any concerns.

rhelmot commented:
The ball is in my court in terms of the plugin initialization. After that I'll merge it.

rhelmot commented:
I squashed and merged this manually. The plugin initialization stuff was pretty easy to fix. Thanks a ton!

rhelmot commented:
@m1ghtym0 can you merge heaphopper? There are conflicts.

m1ghtym0 commented:
@rhelmot Thanks, I've merged and adjusted to your changes.
